riotctrl_shell: initial import of shell interaction spawn

miri64 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
One thing I really like about the test utils in https://github.com/RIOT-OS/Release-Specs/pull/79 are the mixins to activate functionality to interact with a shell command. Based on that I provide `ShellInteraction` class here that is supposed to be used as a base class for these kind of mixins. As an example I provided a shell interaction implementation for `ping6` and `reboot`. They allow for something like

```
    class Pinger(GNRCICMPv6Echo, Reboot):
        pass

    node = riotnode.node.RIOTCtrl(application_directory=`optional_dir_if_not_curr_dir`,
                                  env=node_specific_env_dict)
    pinger = Pinger(node)
    pinger.reboot()
    pinger.ping6("ff02::1")
```
<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
I provided some pytest tests

```sh
virtualenv -p python3 /tmp/env
. /tmp/env/bin/activate
pip install -r dist/pythonlibs/riotctrl_shell/requirements.txt pytest
pytest dist/pythonlibs/riotctrl_shell/
PYTHONPATH=${PWD}/dist/pythonlibs:${PYTHONPATH} pytest dist/pythonlibs/riotctrl_shell/
```

If it fails you might just need to initialize the TAP interface first:

```
sudo ./dist/tools/tapsetup/tapsetup
```

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references
Depends on ~~#10949~~ (now on the currently still private [but visible for maintainers](https://github.com/RIOT-OS/riotctrl/) repo).
<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

cladmi commented:
Without looking at the implementation, I am more thinking about a wrapper around `node` or the `expect` instance than as a base class to `expect`.
I do not really want to go with the inheritance scheme here and do more composition.

    shell =  MyTermClass(node)
    shell.reboot()


One of the reason is it could allow having different wrappers for the same instance. Which is way better for namespacing.

    udp_server = MyUDPServer(node)
    tcp_server = MyTCPServer(node)

    udp_server.listen('1234')
    tcp_server.listen('80')

    udp_server.packet_callback = test_udp_cb
    tcp_server.packet_callback = test_tcp_cb

Otherwise every method must be namespaced `shell.tcp_server_listen()` which feels weird.
And multiple inheritance leads to issues when you cannot really substitute an object for the other anymore.

This still allows to do composed objects with a separate class, that would here be the `mixins`:

    class MySuperClassUnderTest():
        def __init__(self, node):
                self.tcp_server = TCPServerShell(node)
                self.udp_server = UDPServerShell(node)


I am just not sure where to put the `Shell` but I think I would still prefer to use the shell as a composition and have maybe a provided class already if you really want to just inherit from it.

    class BasicShellClass():
        SHELL_CLASS = RiotNodeShell
        def __init__(self, node):
            self.shell = self.SHELL_CLASS(node)
        # maybe with calls being forwarded directly to the shell class to be able to do
        # basicshell.wait_ready()

miri64 commented:
@cladmi like this (see last commits). I'm not sure about the decorator myself though (neither correctness nor appropriateness :sweat_smile:)

miri64 commented on dist/pythonlibs/riotnode/riotnode/shell/gnrc.py:
```diff
@@ -0,0 +1,67 @@
+            res = term.expect([r"\d+ bytes from (?P<source>[0-9a-f:]+): "
+                               r"icmp_seq=(?P<seq>\d+) ttl=(?P<ttl>\d+) "
+                               r"(rssi=(?P<rssi>-?\d+) dBm )?"
+                               r"time=(?P<rtt>\d+.\d+) ms",
```
Note to self: `rtt` can also be omitted (when `packet_size < sizeof(uint32_t)`).

miri64 commented:
Ping? I still think that this can help streamline the tests a lot and automate the release testings.

miri64 commented:
If @cladmi agrees I could take over his commits.

aabadie commented:
> If @cladmi agrees I could take over his commits.

With @fjmolinas, we also have the plan to take over this PR.

miri64 commented:
Needs adaptation for https://github.com/RIOT-OS/riotctrl/

miri64 commented:
Rebased to current master, adapted for https://github.com/RIOT-OS/riotctrl/, and removed dependency of `ShellInteraction` to `TermSpawn` (it made little sense). There might be a benefit to move `ShellInteraction` to https://github.com/RIOT-OS/riotctrl/ too. It is basically just a soft wrapper around RIOTCtrl and does not provide any RIOT-specific code.

miri64 commented:
> One of the reason is it could allow having different wrappers for the same instance. Which is way better for namespacing.
> 
> ```
> udp_server = MyUDPServer(node)
> tcp_server = MyTCPServer(node)
> 
> udp_server.listen('1234')
> tcp_server.listen('80')
> 
> # EDIT: hmm the callback is not configured in python but you can see the naming conflict
> udp_server.packet_callback = test_udp_cb
> tcp_server.packet_callback = test_tcp_cb
> ```
> 
> Otherwise every method must be namespaced `shell.tcp_server_listen()` which feels weird.

Yepp. Using Mixins is now not required anymore with the current approach. It is however still possible, if the test implementor prefers so :-).

miri64 commented:
As discussed offline with @fjmolinas, I reworked `ShellInteraction` to utilize [`replwrap`](https://pexpect.readthedocs.io/en/stable/api/replwrap.html). I also added tests and updated the testing procedures accordingly.

miri64 commented on dist/pythonlibs/riotctrl_shell/base.py:
```diff
@@ -0,0 +1,80 @@
+
+        return wrapper
+
+    def _cmd(self, cmd, timeout=-1, async_=False):
```
Its supposed to be protected (? trying to remember terminology from my second semester OOP course...), so private to the class and all its children.

fjmolinas replied:
So the question is whether there is a a use for it being called outside, do we expect all modules to be exhaustive? Do simple functions need a wrapper, or just running the command suffices?

miri64 replied:
Mhh you are right. I make it public then :-)

miri64 replied:
Done

miri64 commented on dist/pythonlibs/riotctrl_shell/gnrc.py:
```diff
@@ -0,0 +1,154 @@
+            rtts["max"] = float(rtts["max"])
+            self.rtts.update(rtts)
+
+    @ShellInteraction.check_term
+    def ping6(self, hostname, count=3, interval=1000, packet_size=4,
```
Seems like a good idea!

miri64 commented on dist/pythonlibs/riotctrl_shell/base.py:
```diff
@@ -0,0 +1,80 @@
+import pexpect.replwrap
+
+
+class ShellInteraction(object):
```
I made it a habit to add this, but I can't remember why (probably some interoperability problems between py2 and 3?)

miri64 replied:
Fixed.

miri64 commented on dist/pythonlibs/riotctrl_shell/base.py:
```diff
@@ -0,0 +1,80 @@
+        :param  cmd: A shell command as string.
+        """
+        self._start_replwrap()
+        return self.replwrap.run_command(cmd, timeout=-1, async_=False)
```
Oops.

miri64 replied:
Fixed

miri64 commented on dist/pythonlibs/riotctrl_shell/base.py:
```diff
@@ -0,0 +1,80 @@
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
+"""
+Mixins for shell interaction
+
+Defines classes to abstract interactions with RIOT shell commands
+"""
```
Should I just push it there? (it's still not public so I can't do PRs :-/)

fjmolinas replied:
Maybe lets keep it in the PR while we still ping-pong on it, and when PR is "ready" split it out and push.

miri64 commented:
@fjmolinas I did not apply your proposals for restructuring yet. May I squash before that?

fjmolinas commented:
> @fjmolinas I did not apply your proposals for restructuring yet. May I squash before that?

Go ahead!

miri64 commented:
Squashed.

miri64 commented:
Another squash push. I reorganized the commits so the tests are with the introduced feature. This way I have to create way less fixups for the changes to come :-)

miri64 commented:
First, I split the command execution from result parsing (but still provided a convenience method to do both) as proposed in https://github.com/RIOT-OS/RIOT/pull/11406#discussion_r448769088.

miri64 commented:
And addressed https://github.com/RIOT-OS/RIOT/pull/11406#pullrequestreview-441371045.

fjmolinas commented on dist/pythonlibs/riotctrl_shell/base.py:
```diff
@@ -11,8 +11,69 @@
+                    command
+                )
+        if lpe >= 2:
+            raise ValueError(
+                "Inter except found - input was incomplete:\n" +
+                command
+            )
```
If a list is provided would be nice to know what was matched, e.g: an `assert`, `hardfault`, `etc`..

miri64 replied:
Done

fjmolinas commented on dist/pythonlibs/riotctrl_shell/base.py:
```diff
@@ -11,8 +11,69 @@
+            lpe = self._expect_prompt(timeout=timeout,
+                                      inter_expect=inter_expect)
+            res.append(self.child.before)
+            if lpe == 2:
```
```suggestion
            if lpe >= 2:
```

miri64 replied:
And done

miri64 commented:
> Some comments, I'm not sure about the `Inter_expect` name, not the best suggestion on my side.

I guess the poll for a new name is open then ;-).

miri64 commented:
How about: `abort_strs`? It makes clear, that it is a list, since plural, it makes clear it `abort`s the shell interaction and it makes clear, that the members of the list need to be literal strings, not regex patterns.

aabadie commented on dist/pythonlibs/riotctrl_shell/gnrc.py:
```diff
@@ -0,0 +1,189 @@
+# ==== Parsers ====
+
+class GNRCICMPv6EchoParser(ShellInteractionParser):
+    class Results():
```
This nested `Results` looks particularly useless and adds unnecessary complexity. All methods could be attached to `GNRCICMPv6EchoParser`.

miri64 replied:
Ok, will port to simple dictionaries.

miri64 replied:
To keep the check methods of the pktbuf parser however, I kept the Results nested class there, but made it a `dict`-like for consistency.

miri64 commented on dist/pythonlibs/riotctrl_shell/base.py:
```diff
@@ -0,0 +1,184 @@
+
+    :param riotctrl: a RIOTCtrl object
+    """
+    parser_class = ShellInteractionParser
```
> First, I split the command execution from result parsing (**but still provided a convenience method to do both**) as proposed in [#11406 (comment)](https://github.com/RIOT-OS/RIOT/pull/11406#discussion_r448769088).

aabadie replied:
This _convenvience method to do both_ is added in the shell interaction class but doesn't depend on it. This design seems weird. I would rather make the difference more explicit.

You could still wrap the usage of both the interaction and the parser in a function if your prefer:

```
def run_and_parse(*args, **kwargs, parser_args=None):
    si = MyShellInterationClass()
    output = si.run_my_cmd(*args, **kwargs)
   parser = MyShellInteractionParser(**parser_args)
   parser.parse(output)
```

miri64 replied:
I don't like your proposal that much (too static, the parser can't be changed) and I don't have a better idea other than what is already present at hand, so I rather remove the convenience function stuff, and require to call `run_cmd | parse` always for now. We can later add convenience functions and discuss it then.

aabadie commented on dist/pythonlibs/riotctrl_shell/gnrc.py:
```diff
@@ -0,0 +1,189 @@
+# directory for more details.
+
+"""
+Mixins for GNRC-related shell interaction
+
+Defines mixins for GNRC-related shell commands
```
AFAICS, there are no more mixins in this module

miri64 replied:
You can still use them as mixins, but will remove.

aabadie commented on dist/pythonlibs/riotctrl_shell/base.py:
```diff
@@ -0,0 +1,184 @@
+# directory for more details.
+
+"""
+Mixins for shell interaction
```
Same comment about mixins.

miri64 commented:
I talked with @fjmolinas offline a bit and we decided to roll back the `REPLWrapper` extension for now. The use case we were aiming at can always be added later and it is more important to get the ball rolling for this and get this into the release for testing than to cover all bases right from the start. @aabadie do you think that is okay as well?

fjmolinas commented:
> @aabadie do you think that is okay as well?

@aabadie suggested this as well so I think he will be fine with it :)

aabadie commented:
> do you think that is okay as well?

I'm ok for moving the REPLWrapper extension in riotctrl later and remove it from this PR. But I still have other requests :)

miri64 commented:
Addressed comments so far.

miri64 commented:
And removed REPLWrapper extension ;-)

fjmolinas commented:
> Regarding `riotctrl_shell`, I think its Ok but I would remove the ` class Reboot(ShellInteraction)`, I think that with a `class Help(ShellInteraction)` they could be part of a another module that would have only have `reboot`, `help`, `version` interactions. With that I think we can move it to `riotctrl` repo.

Maybe a `sys.py` module to match `sc_sys.c`.

miri64 commented:
> > Regarding `riotctrl_shell`, I think its Ok but I would remove the ` class Reboot(ShellInteraction)`, I think that with a `class Help(ShellInteraction)` they could be part of a another module that would have only have `reboot`, `help`, `version` interactions. With that I think we can move it to `riotctrl` repo.
> 
> Maybe a `sys.py` module to match `sc_sys.c`.

:+1:

miri64 commented:
Done.

fjmolinas commented on dist/pythonlibs/riotctrl_shell/sys.py:
```diff
@@ -0,0 +1,37 @@
+
+
+class Help(ShellInteraction):
+    """Reboot ShellInteraction"""
```
wrong help messages.

miri64 replied:
Fixed

fjmolinas commented:
> Done.

IMO you can split `riotctrl_shell` and push into `riotctrl` now. @aabadie any objections?

aabadie commented on dist/pythonlibs/riotctrl_shell/gnrc.py:
```diff
@@ -0,0 +1,170 @@
+
+
+class GNRCPktbufStatsParser(ShellInteractionParser):
+    class Results(dict):
```
Question: what is the benefit of doing this (nested class which inherits a dict) other than looking overly complex ?

miri64 replied:
The main motivation is having useful methods provided for operations in the shell command people usually don't understand immediately (yes! I _always_ get questions about this when we doing testing):

- determine if the packet buffer is empty
- determine how much of the packet buffer was used in total

As to why it is nested: Except for read operations, this class is not useful outside of the parser class, and the setters should be kept private to its outer class (which arguably, I could move there, as with the ping parser, but they are more related to the nested class, so I put them there).

As to why inheriting dict, see https://github.com/RIOT-OS/RIOT/pull/11406#discussion_r449481667: You asked for having it simple in Ping so when we want methods here _and_ consistency it should also be a dict-like.

miri64 replied:
Alternatively, we could have items `'is_empty'` and `'fullest_capacity'` in the dictionary, but since they are dependent on the values of other items, this would become even more complicated when those are changed and tbh weirder than a nested class inheriting from dict to provide item-dependent methods.

aabadie replied:
But all this could be done without using a nested class anyway. The private methods of Results (used directly from the parser btw), could be private method of the parser.
I mean when things can be done a simpler way, there's no reason to overdesign them.

miri64 replied:
Fine... as I can't convince you, I bow to your will :-P

miri64 commented:
In the last round of commits I

- moved over to [`riotctrl.shell`](https://github.com/RIOT-OS/riotctrl/blob/master/riotctrl/shell.py) and removed `riotctrl_shell.base` and its test.
- simplified and sped up the tests by initializing the RIOTctrl object test-wise (all the special cases for term handling are now tested in `riotctrl` anyways).
- Made `GNRCPktbufStatsParser.Results` a non-nested class `GNRCPktbufStatsResults`, the manipulation of this dict-like items happens in private methods of GNRCPktbufStatsParser, `GNRCPktbufStatsResults` extends dict, to have the `is_empty()` and `fullest_capacity()` methods.

miri64 commented:
May I squash @aabadie @fjmolinas? It's getting out of hand again ^^"

aabadie commented on dist/pythonlibs/riotctrl_shell/tests/test_gnrc.py:
```diff
@@ -0,0 +1,103 @@
+logger = logging.getLogger(__name__)
+
+
+RC = None
```
This machinery needs some clarification. I won't block it now but I don't get the need to call a specific test script from here. This seems reversed to me.

At the beginning, I understood the modules under `riotctrl_shell/tests` as modules for unit testing the interactions/parsers. If it's the case then, it would be better to mock some calls.

miri64 replied:
IMHO mocks are no good here, because then we might realize to late, that the parsing of a specific command does not work anymore, because the command itself changed, so we need to use the real deal here. I never called scripts here, but a specific test application in `native`. `tests/shell` and `tests/gnrc_udp` seemed good target for the specific shell interaction classes.

I was thinking to provide a dedicated application for this (e.g. to not depend on `tap` interfaces for GNRC; I ping `::1` in the only real network application here anyway), but in the end lazyness won ^^"

miri64 replied:
I can add clarification as a comment, but I'm not sure what clarification is needed in your opinion.

miri64 replied:
(I did not change the tests or which application is used with them, I only changed how and when the `riotctrl` is initialized)

aabadie replied:
> we might realize to late, that the parsing of a specific command does not work anymore, because the command itself changed

That's the purpose of test no ? If the output is changing, then it's normal to expect the test that is testing the output will fail.

miri64 replied:
Yes... then I'm confused why you want a mock? Or were you referring to a dedicated application for these tests?

aabadie replied:
I think there's some confusion on both side :)
In fact, I would use riotctlr and expecially the ICMPv6 interaction/parser from the script of applications that requires it. Code that is under `riotctrl_shell/tests` should only be there for testing the internals of the interactions/parsers modules themselves, so maybe they would need some mock (no real terminal is required in fact). For parser, that could just be injecting test strings and verify that the parser does the job correctly.

miri64 replied:
Ahhh okay, yes I can do that ^^"

miri64 replied:
Ok, I wrote now a very simple mock for `RIOTCtrl` to test the functions and used that in the tests and used legitimate results from the `gnrc_udp` test as constant input strings for testing parsing.

aabadie replied:
Thanks, that looks good!

miri64 commented on dist/pythonlibs/riotctrl_shell/tests/common.py:
```diff
@@ -0,0 +1,34 @@
+
+
+def init_ctrl():
+    rc = MockRIOTCtrl("foobar", env={"BOARD": "native"})
```
I could have used [`monkeypatch`](https://docs.pytest.org/en/stable/monkeypatch.html) here, but in the end I would have used `MockRIOTCtrl` here, so no use adding that indirection here.

miri64 commented on dist/pythonlibs/riotctrl_shell/tests/test_gnrc.py:
```diff
@@ -0,0 +1,117 @@
+    assert tap_exists()
+    si = riotctrl_shell.gnrc.GNRCICMPv6Echo(rc)
+    res = si.ping6("::1", interval=100)
+    # mock just returns last input
```
While the comment refers to the mock, since our shell echos the last command as well before the command output, this is also true for RIOT's shell (in case the mock is not used for some reason).

aabadie commented on dist/pythonlibs/riotctrl_shell/tests/test_gnrc.py:
```diff
@@ -0,0 +1,117 @@
+logger = logging.getLogger(__name__)
+
+
+def tap_exists(tap="tap0"):
```
Do you think we could get rid of calling `ip link show tap` ? That would simplify automatic testing.

miri64 replied:
Oops I planned to remove this but forgot ^^"

miri64 replied:
Removed.

aabadie commented on dist/pythonlibs/riotctrl_shell/requirements.txt:
```diff
@@ -0,0 +1 @@
+git+ssh://git@github.com/RIOT-OS/riotctrl
```
```suggestion
git+ssh://git@github.com/RIOT-OS/riotctrl
pytest
flake8
```

miri64 replied:
Shouldn't this rather be a dependency of testing? I know tox can be configured as such.

aabadie replied:
You are right. I edited the tox.ini above. If it works please ignore the change in requirements.txt :)

aabadie commented on dist/pythonlibs/riotctrl_shell/tests/test_sys.py:
```diff
@@ -0,0 +1,33 @@
+
+import riotctrl_shell.sys
+
+from common import init_ctrl
```
```suggestion
from .common import init_ctrl
```

aabadie commented on dist/pythonlibs/riotctrl_shell/tests/test_gnrc.py:
```diff
@@ -0,0 +1,104 @@
+import riotctrl_shell.gnrc
+import subprocess
+
+from common import init_ctrl
```
```suggestion
from .common import init_ctrl
```

miri64 commented:
Applied your latest round of comments and squsahed.

miri64 commented:
Yay :tada:, thanks for the review and the patience with my impatience 😅

fjmolinas commented:
Awesome to see this in!
